src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs

```diff
@@ -865,65 +865,73 @@ private int ReadBuffer(Span<char> userBuffer, out bool readToUserBuffer)
             return task;
         }
 
-        private async Task<string?> ReadLineAsyncInternal()
+        private Task<string?> ReadLineAsyncInternal()
         {
-            if (_charPos == _charLen && (await ReadBufferAsync(CancellationToken.None).ConfigureAwait(false)) == 0)
-            {
-                return null;
-            }
-
-            StringBuilder? sb = null;
+            return _charPos == _charLen
+                ? NoDataInTheBuffer()
+                : ConsumeBufferedData();
 
-            do
+            async Task<string?> NoDataInTheBuffer()
             {
-                char[] tmpCharBuffer = _charBuffer;
-                int tmpCharLen = _charLen;
-                int tmpCharPos = _charPos;
-                int i = tmpCharPos;
+                if (await ReadBufferAsync(CancellationToken.None).ConfigureAwait(false) == 0)
+                    return null;
 
+                return await ConsumeBufferedData().ConfigureAwait(false);
+            }
+            async Task<string?> ConsumeBufferedData()
+            {
+                StringBuilder? sb = null;
                 do
                 {
-                    char ch = tmpCharBuffer[i];
+                    char[] tmpCharBuffer = _charBuffer;
+                    int tmpCharLen = _charLen;
+                    int tmpCharPos = _charPos;
+                    int i = tmpCharPos;
 
-                    // Note the following common line feed chars:
-                    // \n - UNIX   \r\n - DOS   \r - Mac
-                    if (ch == '\r' || ch == '\n')
+                    do
                     {
-                        string s;
+                        char ch = tmpCharBuffer[i];
 
-                        if (sb != null)
-                        {
-                            sb.Append(tmpCharBuffer, tmpCharPos, i - tmpCharPos);
-                            s = sb.ToString();
-                        }
-                        else
+                        // Note the following common line feed chars:
+                        // \n - UNIX   \r\n - DOS   \r - Mac
+                        if (ch is '\r' or '\n')
```

Comment:
This produces identical IL and asm, doesn't it?  I would not expect a perf improvement from this.  What will be different and might yield a micro-improvement is `ch == '\r' | ch == '\n'`, removing a (well-predicted) branch.